### PR TITLE
Periodic mesh in omega_h

### PR DESCRIPTION
**Pull request by joshia5**

### CODE REVIEW

**Comment by cwsmith** [comment]
@joshia5 Will merging this cause any problems with non-periodic meshes?  Was there an issue with osh files not being compatible?

**Comment by joshia5** [comment]
@cwsmith as this pr stands now, i am expecting the merge to overwrite the mixed mesh sim2osh converter which is in scorec/master and create compatibility issues for mixed meshes because this work does not yet support mixed-periodic meshes. However any workflows based on sim2osh
mesh conversion that are using your branch https://github.com/SCOREC/omega_h/tree/cws/meshSimReader, should not be affected

**Comment by cwsmith** [comment]
@joshia5 Thank you.  Is it possible to maintain support for simmetrix-to-osh conversion for all three (periodic, mixed, and mono-topology)?  Ideally, I'd like to merge everything (including https://github.com/SCOREC/omega_h/tree/cws/meshSimReader) into `master`.

**Comment by joshia5** [comment]
@cwsmith there are two ways to go on about it. first one is to make a robust utility which can handle all three types of
meshes. this will require a higher development time. second is to make individual utilities for all three (simplex, mixed and periodic, however the mixed and simplex converters can be grouped together).  from where the code is now, the second option should not be much difficult to implement.

**Comment by cwsmith** [comment]
@joshia5 I think having a converter that supports simplex+mixed will cover 90% of our usage so having a separate converter that deals with periodic is OK.  Adding checks to each converter to ensure that the meshes provided by the user are of the correct type would make this strategy robust.

**Review by cwsmith — CHANGES_REQUESTED** [review]
Looks good. I submitted a small PR to your branch to fix a few small issues:
https://github.com/joshia5/omega_h/pull/2

File: src/Omega_h_remotes.hpp [hunk 1]
```diff
@@ -20,6 +20,16 @@ Remotes identity_remotes(CommPtr comm, LO n);
 Remotes owners_from_globals(
     CommPtr comm, Read<GO> globals, Read<I32> own_ranks);
 
+struct c_Remotes {
```
[the remark this answers is not in the document]
Comment (joshia5) [reply src/Omega_h_remotes.hpp:23]: @cwsmith The  compressed remotes data structure a.k.a `c_Remotes` is used for storing matches

File: src/Omega_h_matchMeshsim.cpp [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+
+    pPList matches = EN_getMatchingEnts(edge, 0, 0);
+    void *iterM = 0;
+    pEdge match;
+    count_matches = 0;
+    if (matches) {
+      assert (PList_size(matches) > 1);
+      while ((match = (pEdge)PList_next(matches, &iterM))) {
+        if (EN_id(match) != EN_id(edge)) {
+          ++count_matches;
```
[the remark this answers is not in the document]
Comment (joshia5) [reply src/Omega_h_matchMeshsim.cpp:300]: I think there might be some confusion. `count_matches` is number of matches for each 'self' entity. `++count_matches` seems ok to me.

[the remark this answers is not in the document]
Comment (cwsmith) [reply src/Omega_h_matchMeshsim.cpp:300]: I should have put this comment one line lower. I'm suggesting replacing:

```
          ++count_matches;
          ++count_matches;
```

with:

```
          count_matches+=2;
```

[the remark this answers is not in the document]
Comment (cwsmith) [reply src/Omega_h_matchMeshsim.cpp:300]: Yeah... I misread this as `count_matches++` twice.  Sorry about that.

File: src/Omega_h_matchMeshsim.cpp [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+    auto model_match = Read<LO>(host_model_matches.write());
+    mesh->set_model_ents(ent_dim, model_ids);
+    mesh->set_model_matches(ent_dim, model_match);
+  }
+  return;
+}
+
+void matchRead(filesystem::path const& mesh_fname, filesystem::path const& mdl_fname,
+    CommPtr comm, Mesh *mesh, I8 is_in) {
+  mesh->set_periodic(1);
```
[the remark this answers is not in the document]
Comment (joshia5) [reply src/Omega_h_matchMeshsim.cpp:467]: > I assume that if someone loads a non-periodic/matched mesh then the converter will fail?
Yes

**Review by cwsmith — APPROVED** [review]

**Pull request merged**